### ISSUE TITLE
DB constraint violations for clients other than <bean class="org.mitre.oauth2.model.ClientDetailsEntity" p:id="1" ...>

### ISSUE THREAD

**Issue by lhoekenga** [uchicago/shibboleth-oidc#50]
This is an evolution of #47 

I am having constraint violations when inserting entries into the client_details table for client definitions other than <bean class="org.mitre.oauth2.model.ClientDetailsEntity" p:id="1" ...>.

I've seen this with both the HSQLDB and MySQL backends.
I think it may be because the SELECT used to find the client_details entry sometimes uses the "p:id" attribute from the ClientDetailsEntity bean, and sometimes uses the auto-incremented ID number from the sql entry.

Starting from newly instantiated tables, the basic flow seems to be...
User 1:
SELECT ... FROM client_details WHERE (id = X)
_where X is the p:id for the matching ClientDetailsEntity_

*empty result set*

INSERT INTO client_details ...

_depending on platform_
HSQLDB:  CALL IDENTITY()
MYSQL: SELECT LAST_INSERT_ID()

SELECT ... FROM client_details WHERE (id = X)
_where X is result returned from that lookup_

Access granted!

**Here's where things go south**
User 2:
SELECT ... FROM client_details WHERE (id = **1**)
_The user is accessing the first OIDC client defined, the select returns an entry and everything is happy._

User 3:
SELECT ... FROM client_details WHERE (id = X)
_where X is the p:id for the matching ClientDetailsEntity_

*empty result set*

INSERT INTO client_details ...

Throw an error
HSQLDB: java.sql.SQLIntegrityConstraintViolationException: integrity constraint violation: unique constraint or index violation; SYS_CT_10124 table: CLIENT_DETAILS
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)

MYSQL: com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Duplicate entry 'client' for key 'client_id'
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)

Access denied.
[bad hsqldb 17oct17.txt](https://github.com/uchicago/shibboleth-oidc/files/1392872/bad.hsqldb.17oct17.txt)
[good mysql 17oct17.txt](https://github.com/uchicago/shibboleth-oidc/files/1392873/good.mysql.17oct17.txt)
[good hsqldb 17oct17.txt](https://github.com/uchicago/shibboleth-oidc/files/1392874/good.hsqldb.17oct17.txt)
[bad mysql 17oct17.txt](https://github.com/uchicago/shibboleth-oidc/files/1392875/bad.mysql.17oct17.txt)

**Comment by lhoekenga** [comment]
I compared the SQL generated by the overlay...
    SELECT ... FROM client_details WHERE (id = ?)

to the SQL generated by the vanilla MitreID Connect.
    SELECT ...  FROM client_details WHERE (client_id = ?)

The SQL from MitreID make a lot more sense in this context.

**Comment by mmoayyed** [comment]
Progress:

1. Fixed issue with client queries.
2. Upgraded to MitreID 1.3.x
3. Upgraded to IdP 3.3.2
4. In the process of updating README
5. Updated schemas to match 1.3.x. In the process of doing so, the oidc.properties file simply now points to a `type` setting rather than a path to schema file since 1.3 sort of explodes the schema into multiple files.

Some notable changes:

1. I have cleaned up the `oidc-protocol-*.xml` files in the `system` directory to be a bit more organized. A number of new beans and such are added or modified as part of switching from 1.2.6 to 1.3.1. For the most part, it does not impact anything at all.
2. In the course of fixing the issue, I resorted to the following compromise which I think is, in fact, an improvement: the code will now use MitreID's default/configured option to manage client definitions. This means definitions are loaded from and saved into the database. This allows you the flexibility to directly run SQL queries to create clients without having to deal with XML and works better as it's more aligned with MitreID's way of doing things. Then, I added a bootsrapper that can take client definitions from XML and import them into the database on startup, and import does a save/update.

Will push soon.
